### PR TITLE
Angular 4 and better typing

### PR DESCRIPTION
**Pull request by BlackLanzer**
Updated to Angular 4 and fixed some compiler errors.

I also added explicit types in functions because most project has --noImplicitAny in compiler settings and until there is not a published release it breaks everything.

### CODE REVIEW

**Comment by jaumard** [comment]
Hi @BlackLanzer ! Thanks for this ! Looks good :) didn't have the time to update it ^^ 
Does it fix this error : 
`[default] /Users/jaumard/LISAProjects/lisa-ui/src/app/pages/home/home.component.ts:105:33 
    Property 'getWidgetById' does not exist on type 'DashboardComponent'` ? I have this one on my current project but didn't find why it's not happy as it works ^^

**Comment by jaumard** [comment]
Also why do you comment tests ? They are irrelevant because it's auto generated test but just want to know if there a reason

**Comment by BlackLanzer** [comment]
I don't see the error about getWidgetById. Are you sure you are on the latest version?

I commented the tests because they don't compile because of some wrong file names or some other things and I didn't feel like fixing it because they don't do anything anyway. Maybe I should have written about it somewhere.

**Comment by jaumard** [comment]
I just try with last version and don't have this error anymore ^^ don't remember fixing it lol 
Ok for the tests ^^ no problem. Did you test the demo see with this modification it still working ?

**Comment by jaumard** [comment]
Hum also does this mean that after merging it will not work anymore for angular 2 ? I didn't update myself about the v4 yet

**Comment by jaumard** [comment]
Just test the demo on your fork; doesn't work anymore, to try just do this : 
```
cd demo
npm i 
npm start 
```
Can you take a look ?

**Comment by BlackLanzer** [comment]
> Hum also does this mean that after merging it will not work anymore for angular 2 ? I didn't update myself about the v4 yet

I think it should work because I didn't add anything new.

About the demo: i can't try the current version because i should uninstall angular-cli, install the old and deprecated version, etc.  
What kind of error does it give to you?  
I can make it work updating package.json with the current versions and adding one line to angular-cli.json, but I don't know what you want to do with the versions. My changes should work with angular 2 AFAIK.  
In my opinion you should at least update angular-cli in the demo because it is a very old version.

File: package.json [hunk 1]
```diff
@@ -25,8 +25,8 @@
   },
   "homepage": "https://github.com/jaumard/ng2-dashboard#readme",
   "dependencies": {
-    "@angular/core": "^2.4.5",
-    "rxjs": "^5.0.3",
-    "zone.js": "^0.7.6"
+    "@angular/core": "^4.0.2",
+    "rxjs": "^5.1.0",
+    "zone.js": "^0.8.4"
```
Comment (BlackLanzer) [inline package.json]: I don't know why i changed this :) It shouldn't be needed

**Comment by jaumard** [comment]
About the demo I just want something who is working whatever the versions of angular or cli, you can update whatever you want and I agree that it will be better to update at least the cli as they are now in v1 stable. 
Because now it just give me : 

```
ERROR in /Users/jaumard/Downloads/ng2-dashboard-master 2/node_modules/@angular/core/src/errors.d.ts (9,33): Initializers are not allowed in ambient contexts.
/Users/jaumard/Downloads/ng2-dashboard-master 2/node_modules/@angular/core/src/errors.d.ts (10,43): Initializers are not allowed in ambient contexts.
/Users/jaumard/Downloads/ng2-dashboard-master 2/node_modules/@angular/core/src/errors.d.ts (11,42): Initializers are not allowed in ambient contexts.
/Users/jaumard/Downloads/ng2-dashboard-master 2/node_modules/@angular/core/src/errors.d.ts (12,43): Initializers are not allowed in ambient contexts.
/Users/jaumard/Downloads/ng2-dashboard-master 2/node_modules/@angular/core/src/errors.d.ts (13,35): Initializers are not allowed in ambient contexts.

ERROR in /Users/jaumard/Downloads/ng2-dashboard-master 2/node_modules/@angular/core/src/animation/dsl.d.ts (34,33): Initializers are not allowed in ambient contexts.

ERROR in /Users/jaumard/Downloads/ng2-dashboard-master 2/node_modules/@angular/core/src/animation/animation_metadata_wrapped.d.ts (12,33): Initializers are not allowed in ambient contexts.

ERROR in /Users/jaumard/Downloads/ng2-dashboard-master 2/demo/node_modules/@types/jasmine/index.d.ts (40,52): '=' expected.
/Users/jaumard/Downloads/ng2-dashboard-master 2/demo/node_modules/@types/jasmine/index.d.ts (42,62): '=' expected.
/Users/jaumard/Downloads/ng2-dashboard-master 2/demo/node_modules/@types/jasmine/index.d.ts (40,38): A parameter initializer is only allowed in a function or constructor implementation.
/Users/jaumard/Downloads/ng2-dashboard-master 2/demo/node_modules/@types/jasmine/index.d.ts (40,46): Cannot find name 'keyof'.
/Users/jaumard/Downloads/ng2-dashboard-master 2/demo/node_modules/@types/jasmine/index.d.ts (42,46): A parameter initializer is only allowed in a function or constructor implementation.
/Users/jaumard/Downloads/ng2-dashboard-master 2/demo/node_modules/@types/jasmine/index.d.ts (42,56): Cannot find name 'keyof'.
/Users/jaumard/Downloads/ng2-dashboard-master 2/demo/node_modules/@types/jasmine/index.d.ts (58,42): Cannot find name 'Partial'.
/Users/jaumard/Downloads/ng2-dashboard-master 2/demo/node_modules/@types/jasmine/index.d.ts (99,22): Cannot find name 'Partial'.
/Users/jaumard/Downloads/ng2-dashboard-master 2/demo/node_modules/@types/jasmine/index.d.ts (99,35): Cannot find name 'Partial'.

ERROR in Error encountered resolving symbol values statically. Calling function 'ɵmakeDecorator', function calls are not supported. Consider replacing the function or lambda with a reference to an exported function, resolving symbol NgModule in /Users/jaumard/Downloads/ng2-dashboard-master 2/node_modules/@angular/core/core.d.ts, resolving symbol Ng2DashboardModule in /Users/jaumard/Downloads/ng2-dashboard-master 2/components/index.ts, resolving symbol Ng2DashboardModule in /Users/jaumard/Downloads/ng2-dashboard-master 2/components/index.ts
```

**Comment by jaumard** [comment]
@BlackLanzer I updated the demo, now all good except this error :

```
ERROR in Error encountered resolving symbol values statically. Calling function 'ɵmakeDecorator', function calls are not supported. Consider replacing the function or lambda with a reference to an exported function, resolving symbol NgModule in /Users/jaumard/TrailsProjects/ng2-dashboardBlackLanzer/node_modules/@angular/core/core.d.ts, resolving symbol Ng2DashboardModule in /Users/jaumard/TrailsProjects/ng2-dashboardBlackLanzer/components/index.ts, resolving symbol Ng2DashboardModule in /Users/jaumard/TrailsProjects/ng2-dashboardBlackLanzer/components/index.ts
webpack: Failed to compile.
``` 

From what I understand it's because this lib is not AOT compatible :( have you any experience with this ? I don't

**Comment by jaumard** [comment]
I'm merging this one like this I can search more easily what is the problem, but I you manage to find don't hesitate to create another PR :) 
Thanks for your work !

**Comment by BlackLanzer** [comment]
That error doesn't make any sense. If you do `ng serve` first it gives the error, but if you save one file to force the ricompilation it compiles without problems. The problem is you can't do `ng build` because it just fails.

This is my first time using angular, so I'm not sure on what's happening. In the project where I'm using your library it's working fine.

**Comment by jaumard** [comment]
Yeah I'm pretty new to angular too :( ^^ 
I notice that the second time it pass when recompile and the project is working but yeah ng build just fail.
I just try to include this branch into my current project and I have the same error :( but if I use master no errors

**Comment by BlackLanzer** [comment]
After many tries I managed to make it build. It needs some more studying, but it works if you move outside the demo folder and put inside the "library" folder (and of course change the imports in app.module.ts, app.component.ts and my-widget.component.ts). The resulting tree is something like this:
```
demo
├───e2e
├───ng2-dashboard
│   ├───components
│   │   ├───dashboard
│   │   └───widget
│   └───directives
└───src
    ├───app
    │   ├───my-widget
    │   └───shared
    ├───assets
    └───environments
```
Let me know if this works for you, but don't ask me why it does xD

**Comment by jaumard** [comment]
Ok same result it work for me too... very strange ^^ I'll try to search and find the problem because it's fix the demo but not for other project that want to use it :) because I still have the error on my project :( 
Thanks for the tips ! It may help to find the why ^^

**Comment by BlackLanzer** [comment]
I started to work on this solution after reading [this](https://github.com/angular/angular/issues/15767#issuecomment-292740420) where it says that angular-cli copies all the files somehwere, so it can break things. Maybe this could help you.

**Comment by jaumard** [comment]
Everything is fixed and I rename/publish the repo to ngx-dashboard :) thanks for your help with this !

**Pull request merged**